sources/web/datalab/polymer/components/datalab-files/datalab-files.ts

```diff
@@ -404,6 +435,70 @@ class FilesElement extends Polymer.Element {
       return Promise.resolve(null);
     }
   }
+
+  /**
+   * Creates a directory picker modal to get the user to choose a destination for the
+   * selected item, sends a copy item API call, then refreshes the file list. This only
+   * works if exactly one item is selected.
+   * TODO: Consider allowing multiple items to be copied.
+   */
+  _copySelectedItem() {
+    const selectedIndices = this.$.files.getSelectedIndices();
+
+    if (selectedIndices.length === 1) {
+      const i = selectedIndices[0];
+      const selectedObject = this._fileList[i];
+
+      const options: DialogOptions = {
+        title: 'Copy Item',
+        okLabel: 'Copy Here',
+        big: true,
+      };
+      Utils.showDialog(DirectoryPickerDialogElement, options)
+        .then((closeResult: DirectoryPickerDialogCloseResult) => {
+          if (closeResult.confirmed) {
+            const newPath = closeResult.directoryPath;
+            return ApiManager.copyItem(selectedObject.path, newPath)
+              .then(() => this._fetchFileList());
+          } else {
+            return Promise.resolve(null);
+          }
+        });
+    }
+  }
+
+  /**
+   * Creates a directory picker modal to get the user to choose a destination for the
+   * selected item, sends a rename item API call, then refreshes the file list. This only
+   * works if exactly one item is selected.
+   * TODO: Consider allowing multiple items to be copied.
+   */
+  _moveSelectedItem() {
+    const selectedIndices = this.$.files.getSelectedIndices();
+
+    if (selectedIndices.length === 1) {
+      const i = selectedIndices[0];
+      const selectedObject = this._fileList[i];
+
+      const options: DialogOptions = {
+        title: 'Move Item',
+        okLabel: 'Move Here',
+        big: true,
+      };
+      Utils.showDialog(DirectoryPickerDialogElement, options)
+        .then((closeResult: DirectoryPickerDialogCloseResult) => {
+          if (closeResult.confirmed) {
+            const newPath = closeResult.directoryPath;
+            // Moving is renaming.
+            return ApiManager.renameItem(selectedObject.path, newPath + '/' + selectedObject.name)
+              .then(() => this._fetchFileList());
```

Comment:
What happens if the rename fails? Should this "then" be a "finally" instead so that it always updates the list?